Undo log split to reduce contention

andreitokar opened the pull request.

andreitokar committed:
Undo log is split into per-transaction maps

andreitokar committed:
Merge remote-tracking branch 'h2database/master' into undo-log-split

katzyn commented on h2/src/test/org/h2/test/store/TestTransactionStore.java:
```diff
@@ -52,7 +52,8 @@ public void test() throws Exception {
         testTransactionAge();
-        testStopWhileCommitting();
+// TODO: figure out why it hangs
+//        testStopWhileCommitting();
```
https://github.com/h2database/h2database/blob/6d22a935c116788e44e245f11e3099d640bff573/h2/src/test/org/h2/test/store/TestTransactionStore.java#L356
https://github.com/h2database/h2database/blob/6d22a935c116788e44e245f11e3099d640bff573/h2/src/test/org/h2/test/store/TestTransactionStore.java#L415-L417
`tx.getStatus()` always `Transaction.STATUS_OPEN` here, so counter is never incremented.

andreitokar replied:
I think i figured this one out, will fix it tonight. There is nothing stored on disk about committing transaction, until it actualy closes, so there is no way that store closure will catch it in “commiting” state. Before this PR, undo log was cleared one entry at a time, and on opening, absence of the first entry was treated as “partially **committed**” transaction. I am going temporarily rename undoLog map in some predictable way instead, to give a hint for opening logic, that transaction was alredy committed.

katzyn commented on h2/src/main/org/h2/mvstore/tx/TransactionMap.java:
```diff
@@ -77,20 +77,29 @@ public long sizeAsLong() {
         BitSet committingTransactions;
         MVMap.RootReference mapRootReference;
-        MVMap.RootReference undoLogRootReference;
+        MVMap.RootReference undoLogRootReferences[];
```
It's better to declare it as `MVMap.RootReference[] undoLogRootReferences`.

katzyn commented on h2/src/main/org/h2/mvstore/tx/TransactionStore.java:
```diff
@@ -50,7 +51,9 @@
      */
-    final MVMap<Long, Object[]> undoLog;
+    @SuppressWarnings("unchecked")
+    final MVMap<Long,Object[]> undoLogs[] = (MVMap<Long,Object[]>[])new MVMap[MAX_OPEN_TRANSACTIONS];
```
This cast is not required.

katzyn commented on h2/src/main/org/h2/mvstore/tx/TransactionStore.java:
```diff
@@ -562,13 +544,22 @@ synchronized void endTransaction(Transaction t, boolean hasChanges) {
+                for (int i = openTrans.nextSetBit(0); empty && i >= 0; i = openTrans.nextSetBit(i + 1)) {
+                    MVMap<Long, Object[]> undoLog = undoLogs[i];
+                    if (undoLog != null) {
+                        empty = undoLog.isEmpty();
```
May be replace it with
```Java
if (undoLog != null && !undoLog.isEmpty) {
    return;
}
```
and remove `empty` variable?

andreitokar replied:
I never liked “return;” from inside control structures, always smells like GOTO to me. :smile:

katzyn commented:
OOME in `TestMultiThread.testConcurrentLobAdd()` is reproducible even if test is launched by itself with
```Java
        config.memory = false;
        config.networked = false;
        config.diskUndo = true;
        config.diskResult = true;
        config.traceLevelFile = 3;
        config.throttle = 1;
        config.cacheType = "SOFT_LRU";
        config.cipher = "AES";
```
and increased testing time in `Thread.sleep()`.

With -Xmx128M I got error with following stack trace:
```
	Caused by: java.lang.OutOfMemoryError: Capacity: 35831808
		at org.h2.mvstore.WriteBuffer.grow(WriteBuffer.java:322)
		at org.h2.mvstore.WriteBuffer.ensureCapacity(WriteBuffer.java:302)
		at org.h2.mvstore.WriteBuffer.put(WriteBuffer.java:164)
		at org.h2.mvstore.type.ObjectDataType$ObjectArrayType.write(ObjectDataType.java:1366)
		at org.h2.mvstore.type.ObjectDataType.write(ObjectDataType.java:127)
		at org.h2.mvstore.type.ObjectDataType.write(ObjectDataType.java:121)
		at org.h2.mvstore.Page$Leaf.writeValues(Page.java:1334)
		at org.h2.mvstore.Page.write(Page.java:704)
		at org.h2.mvstore.Page$Leaf.writeUnsavedRecursive(Page.java:1343)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.MVStore.storeNow(MVStore.java:1197)
		at org.h2.mvstore.MVStore.store(MVStore.java:1092)
		... 18 more
```

With -Xmx512M error is the same:
```
	Caused by: java.lang.OutOfMemoryError: Capacity: 181398528
		at org.h2.mvstore.WriteBuffer.grow(WriteBuffer.java:322)
		at org.h2.mvstore.WriteBuffer.ensureCapacity(WriteBuffer.java:302)
		at org.h2.mvstore.WriteBuffer.put(WriteBuffer.java:164)
		at org.h2.mvstore.type.ObjectDataType$ObjectArrayType.write(ObjectDataType.java:1366)
		at org.h2.mvstore.type.ObjectDataType.write(ObjectDataType.java:127)
		at org.h2.mvstore.type.ObjectDataType.write(ObjectDataType.java:121)
		at org.h2.mvstore.Page$Leaf.writeValues(Page.java:1334)
		at org.h2.mvstore.Page.write(Page.java:704)
		at org.h2.mvstore.Page$Leaf.writeUnsavedRecursive(Page.java:1343)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.MVStore.storeNow(MVStore.java:1197)
		at org.h2.mvstore.MVStore.store(MVStore.java:1092)
		... 18 more
```
It is also reproducible with -Xmx2048M, but I lost the stack trace.

With -Xmx128M and only
```Java
        config.memory = false;
```
such failure is also reproducible:
```
General error: "java.lang.IllegalStateException: java.lang.OutOfMemoryError: Capacity: 23887872 [1.4.197/3]" [50000-197]
		at org.h2.message.DbException.get(DbException.java:167)
		at org.h2.message.DbException.convert(DbException.java:306)
		at org.h2.mvstore.db.MVTableEngine$1.uncaughtException(MVTableEngine.java:89)
		at org.h2.mvstore.MVStore.handleException(MVStore.java:2619)
		at org.h2.mvstore.MVStore.panic(MVStore.java:414)
		at org.h2.mvstore.MVStore.store(MVStore.java:1096)
		at org.h2.mvstore.MVStore.tryCommit(MVStore.java:1048)
		at org.h2.mvstore.MVStore.beforeWrite(MVStore.java:2300)
		at org.h2.mvstore.MVMap.beforeWrite(MVMap.java:919)
		at org.h2.mvstore.MVMap.operate(MVMap.java:1493)
		at org.h2.mvstore.MVMap.put(MVMap.java:146)
		at org.h2.mvstore.MVMap.put(MVMap.java:135)
		at org.h2.mvstore.StreamStore.writeBlock(StreamStore.java:196)
		at org.h2.mvstore.StreamStore.put(StreamStore.java:159)
		at org.h2.mvstore.StreamStore.put(StreamStore.java:104)
		at org.h2.store.LobStorageMap.createLob(LobStorageMap.java:211)
		at org.h2.store.LobStorageMap.createClob(LobStorageMap.java:196)
		at org.h2.jdbc.JdbcConnection.createClob(JdbcConnection.java:1948)
		at org.h2.jdbc.JdbcPreparedStatement.setCharacterStream(JdbcPreparedStatement.java:1165)
		at org.h2.jdbc.JdbcPreparedStatement.setCharacterStream(JdbcPreparedStatement.java:1128)
		at org.h2.jdbc.JdbcPreparedStatement.setCharacterStream(JdbcPreparedStatement.java:1143)
		at org.h2.test.db.TestMultiThread$2.call(TestMultiThread.java:134)
		at org.h2.util.Task.run(Task.java:44)
		at java.base/java.lang.Thread.run(Thread.java:844)
....................
	Caused by: java.lang.OutOfMemoryError: Capacity: 23887872
		at org.h2.mvstore.WriteBuffer.grow(WriteBuffer.java:322)
		at org.h2.mvstore.WriteBuffer.ensureCapacity(WriteBuffer.java:302)
		at org.h2.mvstore.WriteBuffer.put(WriteBuffer.java:164)
		at org.h2.mvstore.type.ObjectDataType$ObjectArrayType.write(ObjectDataType.java:1366)
		at org.h2.mvstore.type.ObjectDataType.write(ObjectDataType.java:127)
		at org.h2.mvstore.type.ObjectDataType.write(ObjectDataType.java:121)
		at org.h2.mvstore.Page$Leaf.writeValues(Page.java:1334)
		at org.h2.mvstore.Page.write(Page.java:704)
		at org.h2.mvstore.Page$Leaf.writeUnsavedRecursive(Page.java:1343)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.Page$NonLeaf.writeUnsavedRecursive(Page.java:1129)
		at org.h2.mvstore.MVStore.storeNow(MVStore.java:1197)
		at org.h2.mvstore.MVStore.store(MVStore.java:1092)
		... 18 more
```
But it is not reproducible so easy with only `config.memory = false` and larger heap size.

This test inserts CLOBs with 10,024 characters with auto-commit in two concurrent threads, nothing special. Something is going wrong, write buffer with 181,398,528 bytes is definitely too large for VM with 512 MB heap size limit.

katzyn commented:
I can't reproduce it with the current `master` branch.

andreitokar commented:
Can’t reproduce OOME. What OS/JDK you are on? Any non-default GC settings?
Never mind, i was testing wrong branch. Easily reproducible now.

grandinj commented on h2/src/main/org/h2/mvstore/tx/TransactionMap.java:
```diff
@@ -77,20 +77,29 @@ public long sizeAsLong() {
+                    undoLogSize += rootReference.root.getTotalCount();
+                }
+            }
         } while(committingTransactions != store.committingTransactions.get() ||
```
looking at these loops, are we sure that we always write to these fields in the order 
    committingTransactions, map.getRoot(), store.openTransactions
?
Because otherwise I don't see how we can be sure we get a combination of these references that are mutually consistent.

andreitokar commented:
As long as each individual undo log is coherent with map root and commitingTransactions bitmask, we should be ok. Undo logs do not have to be coherent between themselves. Why do we care if some transaction makes changes to some unrelated map?

andreitokar commented:
@grandinj, maybe i misunderstood your question, and it is really: how we assure that even if we manage to capture “coherent” state which truly existed at some point in time (although there is no such thing with java memory model), it will represent logically consistent data set? Lets put bitmask aside, because it is totaly independent of data map and undo log, so there is no opportunity for inconsistency. But later two pieces may be captured in a state when undo log entry already present, but data map is not modified yet. That inconsistency has indeed to be dealt with. In this specific instance, checks in lines 146 and 149 ensure that coresponding data map entry exists and it’s operationId matches.

andreitokar committed:
Fix TestTransactionStore failure - Persist fact of transaction's commit ASAP
Address code review comments.

andreitokar commented:
@katzyn, It seem's like change targeted to persist fact of transaction's commit ASAP, also fixed (masked) OOM in TestMultithread.testConcurrentLobAdd(). It's that MVStore.rename() is synchronized - just insertion of empty synchronized(store){} block eliminated OOME. It may be just a slowdown due to synchronization that makes the difference.

andreitokar commented:
Or maybe not... keep failing, digging deeper

grandinj commented:
@andreitokar sorry, I should have been more explicit. I'm not worried about this specific situation, sizeAsLong() is hardly a critical method, I'm just asking about the "loop until coherent" strategy in general. 

I am more interested in the reliablity of the other places where we are now doing this.

I know that "loop until coherent" is a valid strategy, but up until now I thought that it relied on very careful ordering of writes to the relevant variables, but perhaps this is safer because we are mostly talking about Atomic* variables?

andreitokar commented:
@grandinj, speaking about “physical” consistency (java memory model level) we are ok, indeed because all variables involved are volatile. Yes, sometimes people are playing with mix of non-volatile and volitile vars, but this is a slippery slope.

andreitokar committed:
Rollback accidental test duration change

andreitokar committed:
remove syncronized from MVStore.renameMap()
make beforeWrite() to call commit() instead of tryCommit() in case of unsaved memory churn

katzyn commented on h2/src/main/org/h2/mvstore/MVStore.java:
```diff
@@ -1068,8 +1070,16 @@ public long tryCommit() {
+        Thread currentThread = Thread.currentThread();
+        Thread storeThread = currentStoreThread.get();
+        if (currentThread != storeThread) { // to avoid re-entrance
+            currentStoreThread.set(currentThread);
```
If `storeThread == null` concurrent `tryCommit()` or `deregisterVersionUsage()` may change its value between `get()` and `set()`, they are not synchronized like this `commit()` method. It looks like `compareAndSet()` should be used here too instead of `set()` or something else should be changed.

katzyn replied:
Any updates?

katzyn commented:
> As long as each individual undo log is coherent with map root and commitingTransactions bitmask, we should be ok.

If only `committingTransactions` is changed in one thread and another thread thies to get coherent `committingTransactions` at the same time it may not detect that it got newer `committingTransactions`, but older `map.getRoot()`?

andreitokar committed:
cope with unfinish map rename due to database closure

andreitokar commented:
@katzyn, what does it mean "newer" or "older" in this context? The task is to get "coherent" state of two volatile variables meaning state, which they had at some recent point in time (like we read them simultaneously), not necessarily "latest" state.

katzyn commented:
I mean that one field is already modified and another field is not modified yet in some commit operation. Is it OK to use such values together in iterators and other methods?

andreitokar commented:
We have map's root and a copy of bitmask of committed transactions, which were fetched "simultaneously". Now even if map has changed afterwards (our root stays the same) or bitmask changed (we hold a copy) it is irrelevant for our operation, i.e. map traversal. We are operating on a "snapshot", not a latest version of a map. Remember, that both tree and bitmask are copy-on-write structures.

grandinj commented:
@andreitokar I think what @katzyn means is that some other code could modify the root, then pause for a while, then modify the bitmask, so we would pick up references to these things that do not "belong" to each other.
(where "pause for a while" is easy to achieve if the thread is re-scheduled in between operations)

andreitokar commented:
Sure, such scenario is completely possible, but that's ok. What you described is: user modifies map, but do not commit transaction for a while. Do we care? Apart from the fact that map version will not be released and db file would be growing (until commit/rollback), it should be bearable.

katzyn commented on h2/src/main/org/h2/mvstore/tx/TransactionStore.java:
```diff
@@ -50,7 +51,9 @@
      */
-    final MVMap<Long, Object[]> undoLog;
+    @SuppressWarnings("unchecked")
+    final MVMap<Long,Object[]> undoLogs[] = new MVMap[MAX_OPEN_TRANSACTIONS];
```
It's better to declare it as `MVMap<Long,Object[]>[] undoLogs`.

I noticed that unused elements of this array are not cleared. If a lot of sessions were opened and closed and application does not use so many sessions at once later this creates a memory leak.

andreitokar replied:
If element not in use all what it holds is a reference to an empty MVMap, which is probably 200-300 bytes at most. I do not want to reallocate maps for frequently used array elements (with low indexes). High-positioned (infrequently used) elements can be either closed and nullified right away or periodically  swept. This also involves closing the map and related synchronization. Need to come up with some heuristic what is worth keeping, and what's not.

katzyn replied:
OK, a few megabytes for an application that allocated a 50,000 sessions at once is not a big price.

I didn't check the retained sizes of these maps before, they share some fields, so total size is not so large.

andreitokar replied:
They should share a single empty root page

katzyn commented:
Possible concurrent change of `currentStoreThread` in `tryComit()` and `commit()` is probably harmless, but I'm not sure about it. (See review comment in `MVStore.commit()`.)

But it's not easy to trace effects of the same possible concurrent change in `deregisterVersionUsage` and `commit()`, because there are different complicated callers. So from my point of view it's better to resolve this minor issue to be sure that logic is sane and consistent.

katzyn commented:
Also I noticed that `MVStore.versions` can be accesed at the same time from the different threads, but `LinkedList` should not be used concurrently. This is not related direcly with changes in this PR, however.

andreitokar commented:
On unrelated topic (just accidentally came across) : Does anyone now what [this line](https://github.com/h2database/h2database/blob/4c06fbca5b6f0d4f71eddef6cb4c005d3ca25814/h2/src/main/org/h2/index/SpatialTreeIndex.java#L118) is for?

katzyn commented:
It was introduced in 0fd0134fdb313e0c82936e752d6365f9100e060a.

It looks like this line is not correct. See its current initialization:
https://github.com/h2database/h2database/blob/4c06fbca5b6f0d4f71eddef6cb4c005d3ca25814/h2/src/main/org/h2/index/SpatialTreeIndex.java#L94-L105

It should be
```Java
if (!isPersistent) {
    store.close();
}
```

katzyn commented:
Or may be something like
```Java
if (!isPersistent || session.getDatabase().getPageStore() == null) {
    store.close();
}
```
Its initialization is very weird for PageStore mode.

Edited. It looks like both PageStore and MVStore backends are closed together when database is closed, so only `!isPersistent` should be considered here.

katzyn commented:
There are a lot of checks for MVStore mode with `database.getMvStore() != null`. Persistent `SpatialTreeIndex` should break a lot of things in PageStore mode. ~~It is not very usable in MVStore mode too, because such indexes are not supported with MVCC that is default for MVStore mode.~~ In MVStore mode code paths are different so spatial indexes can be used with MVCC.

katzyn commented:
Oh, I got it. `SpatialTreeIndex` is PageStore-only, that's why `store.close();` is called unconditionally. But if additional MVStore is shared between different indexes there may be a problem and it also may broke other code paths because `Database.getMvStore() != null` is not expected in PageStore mode.

andreitokar committed:
replace synchronized(this) in MVStore with ReentrantLock

andreitokar commented:
Ok, let's leave alone SpatialTreeIndex for now (unless there are some volunteers to fix it :smile:), it was a side note anyway.
@katzyn, you are absolutely right about currentStoreThread - there is a window of opportunity in  tryCommit() vs. commit() race, and even worse - I totally missed deregisterVersionUsage().
I am replacing synchronized(this) in MVStore for ReentrantLock, and removing currentStoreThread field altogether.
As far as access to MVStore.versions: it is always protected by storeLock (was currentStoreThread). Let me know (hit me with something heavy), if you'll find unprotected code path.

grandinj commented:
SpatialTreeIndex uses an MVStore even though it's a PageStore only thing. It's a little bit of a hack.
Each SpatialTreeIndex allocates it's own Map in the MVStore so there is no sharing.

andreitokar committed:
fix deadlock between compact from background and database closure in foreground
Preserve original exception in CreateTable and AlterTableAddConstraint.

andreitokar committed:
Merge remote-tracking branch 'h2database/master' into undo-log-split

andreitokar committed:
fix missed lock release

katzyn commented:
I don't see paths without acquired `storeLock` that use `MVStore.versions`.

katzyn commented:
But what is that?
https://travis-ci.org/h2database/h2database/jobs/390229557#L4365

Some memory was not released?

I also found an unrelated problem. This test changes value of `config.memory` into `true`.
https://github.com/h2database/h2database/blob/4c06fbca5b6f0d4f71eddef6cb4c005d3ca25814/h2/src/test/org/h2/test/synth/TestRandomSQL.java#L33-L43

That's why failed testing mode was executed with `Test mvStore memory mvcc traceLevelFile defrag` instead of expected `Test mvStore mvcc traceLevelFile defrag`.

katzyn commented:
I can't reproduce such OOME on my system.

andreitokar committed:
move transaction recovery procedure to earlier stage

andreitokar committed:
few changes related to starvation

andreitokar committed:
replace faulty assertion

andreitokar committed:
fix tests

andreitokar committed:
eliminate COMMITTED transaction status

andreitokar committed:
rename COMMITTING -> COMMITTED

andreitokar commented:
On my quad core with latest Fedore and JDK 9 it was almost 100% reproducible, but gone after those latest changes.
It was not really a memory leak. CPU starvation due to tight test loops running compact() was preventing store commit to be executed and resulted in excessive amount of unsaved changes to be accumulated.

andreitokar committed:
removed config.memory = true;

andreitokar committed:
adjust test

andreitokar committed:
Temporarily put back erroneous config settings

This is bad, it opens up the whole new can of warms and need to be dealt with separately

andreitokar committed:
reduce garbage creation

katzyn commented:
Travis builds can be easily restarted without reopening of the PR (if you did it to run tests again and not by mistake). Previous build was successful, let's see what will happen in the restarted build.

Unfortunately even the current master branch has random failures.

katzyn commented:
Perhaps undo log related failures after forced restart should be handled more gracefully?
```
Caused by: java.lang.IllegalArgumentException: A map named undoLog-1 already exists [1.4.197/0]
	at org.h2.mvstore.DataUtils.newIllegalArgumentException(DataUtils.java:854)
	at org.h2.mvstore.DataUtils.checkArgument(DataUtils.java:841)
	at org.h2.mvstore.MVStore.renameMap(MVStore.java:2571)
	at org.h2.mvstore.tx.TransactionStore.commit(TransactionStore.java:426) 
```

andreitokar commented:
Unfortunately, master has even non-random ones, if that misconfiguration in TestRandomSQL is fixed.
As far as Travis restart goes - i was either blind, not signed in or used unsupported device to miss that button, thank you @katzyn.

andreitokar commented:
There should be no errors of this kind. I thought I ironed it out, apparently not... 😕
I would totally expect another one about missing row in primary index, when last store operation before abrupt shutdown happens during row update in meta table.
IMHO, we should make all such problems during startup non-fatal and bring up store in R/O mode instead.
Anyway, at this point i am inclinned to merge it.

katzyn commented:
From my point of view changes are in more or less consistent state and this PR can be merged as is.

But we need to investigate what to do in case of such possible failures on unclean restart later.
